### PR TITLE
feat: Common functions for interfacing with python-irodsclient

### LINKED ISSUE #200: Switch from icommand usage to the python irods client
**Is your feature request related to a problem? Please describe.**
Currently most cubi-tk functions that pull or push data to or from irods/Sodar rely on icommands that are accessed via subprocess and must be available in the current environment. These commands are not part of the stated (or easily/automatically installable) requirements for cubi-tk. The python irods client (which is already used by cubi-tk for other features) also allows transfer of files natively without the need to rely on 'external' dependencies and should be used instead.

**Describe the solution you'd like**
- Implementation of base classes &/ functions that allow file transfer with the native irods client
- change all current references to icommands to use these functions instead

**Describe alternatives you've considered**
NA / status quo

**Additional context**
NA

### PR DESCRIPTION
**Pull request by sellth**
fixes #200 

- [x] add new class iRODSTransfer
- [x] put files to iRODS
- [x] trigger remote checksum computation
- [x] fix tests
- [x] also get files from iRODS
- [x] create common class with utility functions
- [x] use iRODS session manager implemented in irods.check
- [x] create sub-collections (recursive put)
- [x] rsync-like functionality (--sync?) which skips files present on remote
- [x] file progress as (n / n_total)
- [x] check if session is properly handled by contextmanager
- [x] handle `CAT_PASSWORD_EXPIRED` for long transfers

Note: Uploads cannot be cancelled by KeyboardInterrupt (caused by multiprocessing in irods client)

### CODE REVIEW

**Review by Nicolai-vKuegelgen — COMMENTED** [review]
Looks very good overall only a few minor comments.

Would be good if someone who understands mock testing in python better than me could review that part.

File: cubi_tk/irods_common.py [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+        try:
+            self._init_irods().server_version
+            return 0
+        except NonAnonymousLoginWithoutPassword as e:  # pragma: no cover
+            logger.info(self.get_irods_error(e))
+            pass
+        except CAT_INVALID_AUTHENTICATION:  # pragma: no cover
+            logger.warning("Problem with your session token.")
+            self.irods_env_path.parent.joinpath(".irodsA").unlink()
+            pass
```
[the remark this answers is not in the document]
Comment (sellth) [reply cubi_tk/irods_common.py:84]: No, if the file does not exist a password kwarg needs to be provided or `NonAnonymousLoginWithoutPassword` is thrown.

File: cubi_tk/irods_common.py [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+                    sys.exit(1)
+                finally:
+                    self.session.cleanup()
+            t.clear()
+
+    def chksum(self):
+        """Compute remote md5 checksums for all jobs."""
+        common_prefix = os.path.commonpath(self.destinations)
+        for job in self.jobs:
+            if not job.path_src.endswith(".md5"):
```
[the remark this answers is not in the document]
Comment (sellth) [reply cubi_tk/irods_common.py:194]: .md5 files need no server-side checksum. Summoning @mikkonie to confirm

**Comment by sellth** [comment]
I could dress up my "SODAR ingest anything" function as a pilot use case, but maybe we want to separate those PRs?

**Pull request merged**